### PR TITLE
s3: stop rewriting a backslash in an object key to a forward slash

### PR DESCRIPTION
**Pull request by robobun**
S3 object keys are raw byte strings: `\` is legal and names a different object than `/` does. `Bun.S3Client` rewrites it, so two distinct keys silently alias onto one object. The write succeeds, signed and accepted, and nothing reports a problem.

## Repro

```js
const s3 = new Bun.S3Client({ accessKeyId: "key", secretAccessKey: "secret", bucket: "b", endpoint });

await s3.write("a\\b", "hello"); // PUT /b/a/b
await s3.write("a/b", "hello");  // PUT /b/a/b  <- same object

s3.presign("a\\b"); // http://endpoint/b/a/b
s3.presign("a/b");  // http://endpoint/b/a/b
```

## Cause

`encode_uri_component`, which builds both the canonical path that gets signed and the request URL, substitutes the byte instead of percent-encoding it:

```rust
if !ENCODE_SLASH && (c == b'/' || c == b'\\') {
    buffer[written] = if c == b'\\' { b'/' } else { c };
```

Because the signature is computed over the already-rewritten path, the server sees a well-formed, correctly-signed request for the wrong key. There is no `SignatureDoesNotMatch` to surface it.

The same encoder is used for list-objects' `prefix` / `start-after` with `ENCODE_SLASH = true`, where `\` *is* percent-encoded to `%5C`. So `write("a\b")` and `list({ prefix: "a\b" })` already disagreed about what the key is.

## Fix

Percent-encode `\` like any other reserved byte, so it reaches the wire as `%5C`. This is what the AWS SDKs send.

Two things hang off that encoder and are handled in the same change:

- The virtual-hosted bucket check relied on the rewrite: a `\` collapsed to `/`, which a single `contains(b'/')` test then caught. Without the rewrite a crafted bucket could carry a `\` into the host, so the check now tests for both separators on the raw bucket name, before encoding. `"evil.example.com\other"` is still rejected with `ERR_S3_INVALID_ENDPOINT`.
- The key buffer was sized at 1024 bytes, the raw key limit, but percent-encoding expands up to 3x, so the old buffer rejected a 500-byte key of spaces with `ERR_S3_INVALID_PATH` while accepting a 1024-byte key of letters. A backslash now contributes to that expansion, so the buffer is sized for the worst-case encoded length and the 1024-byte S3 key limit is enforced on the raw key instead of emerging from the buffer's capacity. The limit no longer slides with how much of the key percent-encodes.
- The canonical request embeds that key, so its scratch buffer is sized from the same limits. Left at a fixed 4096 bytes, a near-maximal key plus a session token overflowed it and surfaced as `ERR_S3_INVALID_SIGNATURE` (internally `NoSpaceLeft`). `host` and `extra_path` are both slices of the endpoint and partition it, so a single endpoint budget covers both, and the encoded session token is capped at 2048 by `token_encoded_buffer`. The `search_params` overflow on the list-objects path (`sign_request::<true>`, where the query is built from unbounded user prefixes) is untouched and pre-existing; #32973 fixes that properly by heap-allocating.

`\` still works as the bucket/key separator when the bucket is taken from the path (`"bucket\key"`), and `normalize_name` still trims a leading or trailing one. Only the in-key rewrite goes away.

## Verification

`test/js/bun/s3/s3-key-encoding.test.ts`: `write`/`delete`/`presign` with `a\b` vs `a/b` against a stub that records the request line, a 500-byte key that needs the larger buffer, and both rejection cases for a bucket name containing a separator. The backslash and presign assertions fail on `main` (`PUT /bucket/a/b` twice) and pass with the fix.

The existing credentialed round-trip tests in `s3.test.ts` ("should allow backslashes in the path", the leading/trailing separator cases) write and read back the same key, so they are unaffected.

One cross-platform wrinkle the test has to account for: every S3 key argument is parsed as a path-like, and that parser rejects anything over `PATH_MAX` with `ENAMETOOLONG` before the signer runs. macOS sets `PATH_MAX` to 1024, exactly the size S3 allows for a key, so an over-long key is caught a layer earlier there than on Linux (`PATH_MAX` 4096). The key-length rows branch on that rather than asserting the Linux error code everywhere. No user-visible difference, since the two limits coincide.

### CODE REVIEW

File: src/s3_signing/credentials.rs [hunk 1]
```diff
@@ -365,15 +365,24 @@ impl S3Credentials {
...
-        let mut bucket_buffer = [0u8; 63];
+        // A 1024-byte key can percent-encode to 3x that. Bucket names are unreserved
+        // characters only, so their 63-byte limit is also their encoded bound.
+        const MAX_BUCKET_LEN: usize = 63;
+        const MAX_ENCODED_KEY_LEN: usize = 1024 * 3;
+        const MAX_ENDPOINT_LEN: usize = 2048;
+        let mut normalized_path_buffer =
+            [0u8; MAX_ENDPOINT_LEN + MAX_BUCKET_LEN + MAX_ENCODED_KEY_LEN + 2];
+        let mut path_buffer = [0u8; MAX_ENCODED_KEY_LEN];
+        let mut bucket_buffer = [0u8; MAX_BUCKET_LEN];
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/s3_signing/credentials.rs:390]: Confirmed, and the numbers line up. On this branch, a 1024-byte key of spaces (3072 encoded) presigns fine with a 700-byte session token and fails with `ERR_S3_INVALID_SIGNATURE` (that is `NoSpaceLeft`) at 900. On `main` the same key is rejected earlier with `ERR_S3_INVALID_PATH` regardless of the token, so nothing that signed before stops signing.

I am leaving `tmp_buffer` alone here, though. The canonical request also embeds `search_params` verbatim, which is user-supplied, and the session token, which in practice runs past 4 KB. So no fixed constant makes that buffer sufficient, and picking a larger one just moves the wall. #32973 replaces it with a heap allocation in the same function, which is the actual fix; doing it here would conflict with that for no behavioral gain.

Noted the remaining bottleneck in the PR description so the "worst-case encoded length" claim doesn't read as end-to-end.

File: test/js/bun/s3/s3-key-encoding.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      accessKeyId: "key",
+      secretAccessKey: "secret",
+      bucket,
+      region: "us-east-1",
+      virtualHostedStyle: true,
+    });
+
+    expect(() => s3.presign("object.txt")).toThrow(expect.objectContaining({ code: "ERR_S3_INVALID_ENDPOINT" }));
+  },
+);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/s3/s3-key-encoding.test.ts:142]: Leaving these where they are. `test/js/bun/s3/s3.test.ts` is the credentialed suite: it reads `getSecret("S3_R2_*")` and spins up MinIO through docker-compose, so everything in it is gated on credentials being present. These cases need neither, and the directory already holds 13 focused files for exactly that reason (`s3-insecure`, `s3-fd-validation`, `s3-list-encode-overflow`, `s3-queueSize-validation`, `s3-requester-pays`, `s3-storage-class`, ...). Following the local pattern keeps them running on every CI lane.

File: src/s3_signing/credentials.rs [hunk 3]
```diff
@@ -365,15 +365,29 @@ impl S3Credentials {
...
+        const MAX_KEY_LEN: usize = 1024;
+        const MAX_BUCKET_LEN: usize = 63;
+        const MAX_ENCODED_KEY_LEN: usize = MAX_KEY_LEN * 3;
+        const MAX_ENDPOINT_LEN: usize = 2048;
+        if path.len() > MAX_KEY_LEN {
+            return Err(SignError::InvalidPath);
+        }
+        let mut normalized_path_buffer =
+            [0u8; MAX_ENDPOINT_LEN + MAX_BUCKET_LEN + MAX_ENCODED_KEY_LEN + 2];
+        let mut path_buffer = [0u8; MAX_ENCODED_KEY_LEN];
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/s3_signing/credentials.rs:389]: Right, and I was wrong to wave this off earlier. Done in 7f8f830.

I'd argued no constant bounds this buffer because it also holds `search_params`. That's true for `sign_request::<true>` (list-objects builds an unbounded query from user prefixes), but it isn't true for the key-bearing path, which is the one my change widened. There the bound holds: `host` and `extra_path` are both slices of `self.endpoint` and partition it, so a single `MAX_ENDPOINT_LEN` covers both, the encoded session token is capped at 2048 by `token_encoded_buffer`, and the remaining headers are individually bounded. `MAX_ENDPOINT_LEN + MAX_BUCKET_LEN + MAX_ENCODED_KEY_LEN + 4096` is your suggested formula and it fits.

The failing case is now a test: a 1024-byte key of spaces plus a 2000-byte session token presigns. It fails on the parent commit with `ERR_S3_INVALID_SIGNATURE` (that's `NoSpaceLeft`, a spectacularly misleading error) and passes with the buffer sized off the constants.

The list-objects `search_params` overflow is still there, still pre-existing, and still what #32973 fixes properly by heap-allocating.

File: test/js/bun/s3/s3-key-encoding.test.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  const s3 = new S3Client({
+    accessKeyId: "key",
+    secretAccessKey: "secret",
+    bucket: "bucket",
+    endpoint: "http://s3.example.com",
+  });
+  const key = Buffer.alloc(length, fill).toString();
+
+  if (accepted) {
+    expect(s3.presign(key)).toStartWith("http://s3.example.com/bucket/");
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/s3/s3-key-encoding.test.ts]: Tightened in 7f8f830. The boundary cases now compare the full signed path against the key rebuilt byte by byte (`"a".repeat(1024)` vs `"%20".repeat(1024)`), so truncation or omission fails. I kept `.split("?")[0]` rather than `new URL(...).pathname`, to match the assertion style two tests up, and turned the accepted/rejected boolean into the expected per-byte encoding so the case table also states what each byte should become.

**Comment by robobun** [comment]
## Status: diff is green, one red lane is a Buildkite infra flake

[Build 68188](https://buildkite.com/bun/bun/builds/68188) at `109cc5ab`: **285 passed, 1 failed**.

The one failure is `:darwin: 26 aarch64 - test-bun`, which died before executing a single test:

```
Error: buildkite-agent artifact download timed out after 120s for step 'darwin-aarch64-build-bun'.
Refusing to continue with a partial download (would silently fall back to the wrong binary).
```

The sibling `darwin 26 aarch64` shard passed, as did both `darwin 14 aarch64` and both `darwin 14 x64` shards. The build produced no test annotation for anything in this diff (the only test annotation is `test/bake/dev-and-prod.test.ts` on Windows 2019, which passed on retry and is unrelated). The same artifact-download timeout hit this branch once before in [build 68160](https://buildkite.com/bun/bun/builds/68160).

The darwin lanes did earn their keep on the previous push: they caught a real cross-platform bug in the test, fixed in 109cc5ab. Every S3 key argument is parsed as a path-like, and that parser rejects anything over `PATH_MAX` with `ENAMETOOLONG` before the signer runs. macOS sets `PATH_MAX` to 1024, the same size S3 allows for a key, so an over-long key is caught a layer earlier there than on Linux (`PATH_MAX` 4096). The key-length rows now branch on that. Those darwin shards pass on this build.

I have already used my one CI re-roll on this branch (on the same artifact-download flake), so I am not pushing another no-op commit. This needs a maintainer to retry the job or merge.

**Comment by robobun** [comment]
Closing as part of a cleanup of stale pull requests. This PR has had no new commits since 2026-07-03, it conflicts with main, and its last CI run failed. This is not a judgment on the fix itself. If the problem still reproduces on a current build, reopen this PR after a rebase or open a new one against main.
